File: python/56.Merge Intervals.py
# Definition for an interval.
class Interval:
    def __init__(self, s=0, e=0):
        self.start = s
        self.end = e

class Solution:
    def merge(self, intervals):
        """
        :type intervals: List[Interval]
        :rtype: List[Interval]
        """
        if len(intervals) <= 1:
            return intervals

        ret = []
        intervals = sorted(intervals, key = lambda x: x.start)
        tmp_interval = Interval(intervals[0].start, intervals[0].end)

        for i in range(1, len(intervals)):
            this_start = intervals[i].start
            this_end = intervals[i].end
            if this_start <= tmp_interval:
                tmp_interval.end = this_end
                if i == len(intervals) - 1:
                    ret.append(tmp_interval)
            else:
                ret.append(tmp_interval)
                tmp_interval = Interval(intervals[i].start, intervals[i].end)

        return ret
            




if __name__ == '__main__':
    intervals = [Interval(1, 3), Interval(0, 6), Interval(8, 10), Interval(15, 18)]
    print(Solution().merge(intervals))

File: cpp/56. Merge Intervals.cpp
#include <vector>
using namespace std;

class Solution {
public:
    vector<vector<int>> merge(vector<vector<int>>& intervals) {
        sort(intervals.begin(), intervals.end());
        vector<vector<int>>ret;
        int i = 0;
        while(i < intervals.size()){
            vector<int>tmp_vec;
            tmp_vec.push_back(intervals[i][0]);
            int j = i+1;
            int end = intervals[i][1];
            while(j < intervals.size() && intervals[j][0] <= end){
                if (intervals[j][1] > end)
                    end = intervals[j][1];
                ++j;
            }
            tmp_vec.push_back(end);
            ret.push_back(tmp_vec);
            i = j;
        }
        return ret;
    }
};
